torch/csrc/api/include/torch/data/datasets/chunk.h

```diff
@@ -0,0 +1,432 @@
+  using UnwrappedBatchType = typename ChunkReader::BatchType;
+  using BatchRequestType = size_t;
+  using ChunkSamplerType = ChunkSampler;
+  using ExampleSamplerType = ExampleSampler;
+
+  ChunkDataSet(
+      ChunkReader chunk_reader,
+      ChunkSampler chunk_sampler,
+      ExampleSampler example_sampler,
+      size_t preloader_count,
+      size_t batch_size,
+      bool ignore_empty_chunk = false,
+      size_t cache_size = 500)
+      : chunk_reader_(std::move(chunk_reader)),
+        chunk_sampler_(std::move(chunk_sampler)),
+        example_sampler_(std::move(example_sampler)),
+        preloader_count_(preloader_count),
+        batch_size_(batch_size),
+        ignore_empty_chunk_(ignore_empty_chunk),
+        cache_size_(cache_size) {
+    if (preloader_count_ == 0) {
+      throw std::runtime_error(
+          "Preloader count is 0. At least one preloader needs to be specified.");
+    }
+
+    if (batch_size == 0) {
+      throw std::runtime_error(
+          "Batch size is 0. A positive batch size needs to be specified.");
+    }
+
+    if (cache_size_ == 0) {
+      throw std::runtime_error(
+          "Cache size is 0. A positive cache size needs to be specified.");
+    }
+
+    if (cache_size < batch_size){
+      throw std::runtime_error(
+          "Cache size is less than batch size. Cache needs to be large enough to hold at least one batch.");
+    }
+    // chunk_sampler_ =
+    //     std::make_unique<samplers::ThreadSafeSampler<ChunkSamplerType>>(
+    //         std::move(chunk_sampler));
+  }
+
+  virtual ~ChunkDataSet() {
+    free_workers();
+  }
+
+  /// Default get_batch method of BatchDataSet. This method returns
+  /// Example batches created from the preloaded chunks. The implemenation
+  /// is dataset agnostic and does not need overriding in different chunk
+  /// data sets.
+  BatchType get_batch(size_t batch_size) override {
+    if (chunk_buffer_ == nullptr) {
+      throw std::runtime_error(
+          "Dataset needs to call reset() before calling get_batch().");
+    }
+    if (batch_size != batch_size_) {
+      std::string error =
+          "The requested batch size does not match with the initialized batch size.\n"
+          " The requested batch size is " + std::to_string(batch_size) +
+          ", while the data set is created with batch size equal to " +
+          std::to_string(batch_size_);
+
+      throw std::runtime_error(error);
+    }
+    return chunk_buffer_->get_batch(batch_size);
+  }
+
+  /// This will clear any internal state and starts the internal prefetching
+  /// mechanism for the chunk dataset.
+  virtual void reset() {
+    // free workers from previous reset if there is any.
+    free_workers();
+    preload_threads_.clear();
+
+    chunk_reader_.reset();
+
+    size_t chunks_to_load = chunk_reader_.get_chunk_count();
+    chunk_sampler_.reset(chunks_to_load);
+
+    // Creates a new chunk buffer each time we reset the dataset.
```

Comment:
This comment doesn't really add anything new to the code (it's clear you're recreating the buffer). The interesting question here is *why*, and that's something you should explain instead.

Reply:
On top of this reset method, it has a comment explaining that reset will clear any internal state and start from a clean slate. Clear the buffer is one part of it. And by 'clear', we throw away existing one. I have updated the comment to be more specific on this.